sklearn/utils/_testing.py

```diff
@@ -87,6 +87,14 @@
 assert_raises_regexp = assert_raises_regex
 
 
+DTYPE_RELATIVES_TOLERANCES = {
+    np.float32: 1e-4,
+    np.dtype("float32"): 1e-4,
```

Comment:
How does np.dtype("float32") appear ? is it different from np.float32 ?

Reply:
`np.dtype("float32")` is the attribute of `np.array([0.], dtype=np.float32)`. It is not physically the same as the `np.float32` singleton (`is not`) but it is equal to it (`==`).